crates/oxc_parser/src/lexer/keyword_generated.rs

```diff
@@ -0,0 +1,565 @@
+use crate::lexer::Kind;
+#[allow(clippy::enum_glob_use)]
+use crate::lexer::Kind::*;
+
+mod hash {
+    /// Concatenate the first 2 bytes and the last 2 bytes of a slice into a single
+    /// 32-bit integer as an efficient hashing function of JS keywords.
+    /// This approach is backed by the observation that the byte-sequence formed by taking
+    /// the first 2 and last 2 bytes of JS keywords are unique.
+    ///
+    /// SAFETY:
+    /// key.len() >= 2
+    #[inline]
+    pub unsafe fn extract_first_and_last_two_bytes(key: &[u8]) -> u32 {
+        unsafe fn read_u16(input: &[u8]) -> u16 {
+            u16::from(*input.get_unchecked(0)) << 8 | u16::from(*input.get_unchecked(1))
+        }
+        // read first 2 bytes in a u16
+        let first = read_u16(key);
+        let last_bytes = &key[key.len() - 2..];
+        let last = read_u16(last_bytes);
+        u32::from(first) | u32::from(last) << 16
+    }
+
+    /// Use a multiplicative linear congruential generator (MLCG) to map the hash value
+    /// uniformly. Different values of seed are tested to find one with no collisions.
+    #[inline]
+    pub fn hash_u32(input: u32, seed: u32) -> u32 {
+        // The Magic number is taken from
+        // https://www.ams.org/journals/mcom/1999-68-225/S0025-5718-99-00996-5/S0025-5718-99-00996-5.pdf
+        const MAGIC: u64 = 887_987_685;
+        let hash = input ^ seed;
+        ((u64::from(hash) * MAGIC) >> 32) as u32
+    }
+}
+
+const HASH_TABLE_SIZE: usize = 512usize;
+const HASH_TABLE_SEED: u32 = 7_484_039u32;
+static KEYWORD_HASH_TABLE: [(Kind, &'static str); HASH_TABLE_SIZE] = [
```

Comment:
Two questions:

1. Is `&'static str` still needed it here?
2. The table is so big! I wonder if loading this table in and out of cpu cache out weighs every things.

Reply:
Interesting thought.
On the one hand, insofar as the hash function is poor (far from uniform distribution / high collisions), it's _possible_ whole cache line(s) of it are never touched.
On the other hand, it only takes a single entry hit to pay the price.

512 does seem unnecessarily large.
You can try increasing # of attempts, and see if you get a 128 or 256 size table (in a timely manner).